src/icp_canister_backend/src/lib.rs

```diff
@@ -477,21 +583,58 @@ fn add_deposit(
 fn process_episodes() {
     let current_episode = get_current_episode();
     let last_processed_episode = get_last_processed_episode();
+    let current_time = ic_cdk::api::time() / 1_000_000_000;
+    let last_updated_time = LAST_TIME_UPDATED.with(|cell| cell.borrow().get().clone());
+    
+    if current_time == last_updated_time {
+        return;
+    }
 
     let mut total_assets_to_subtract = Nat::from(0u64);
     let mut total_shares_to_subtract = Nat::from(0u64);
+    let mut updated_rewards_at = last_updated_time;
 
     EPISODES.with(|episodes| {
-        let episodes_ref = episodes.borrow();
-
-        for episode_id in (last_processed_episode)..current_episode {
-            if let Some(episode) = episodes_ref.get(&episode_id) {
+        let mut episodes_ref = episodes.borrow_mut();
+        
+        for episode_id in last_processed_episode..current_episode {
+            if let Some(mut episode) = episodes_ref.get(&episode_id) {
+                let episode_finish_time = (episode_id + 1) * EPISODE_DURATION;
+                
+                let reward_rate_contribution = reward_rate_per_share(updated_rewards_at, episode_finish_time);
+                ACCUMULATED_REWARD_PER_SHARE.with(|cell| {
+                    let current_acc = cell.borrow().get().clone().0;
+                    cell.borrow_mut().set(StorableNat(current_acc + reward_rate_contribution)).ok();
+                });
+                
+                updated_rewards_at = episode_finish_time;
+                
+                POOL_REWARD_RATE.with(|cell| {
+                    let current_rate = cell.borrow().get().clone().0;
+                    if current_rate >= episode.reward_decrease {
+                        cell.borrow_mut().set(StorableNat(current_rate - episode.reward_decrease.clone())).ok();
+                    } else {
+                        cell.borrow_mut().set(StorableNat(Nat::from(0u64))).ok();
+                    }
+                });
+                
+                episode.acc_reward_per_share_on_expire = ACCUMULATED_REWARD_PER_SHARE.with(|cell| cell.borrow().get().clone().0);
+                episodes_ref.insert(episode_id, episode.clone());
+                
                 total_assets_to_subtract += episode.assets_staked.clone();
                 total_shares_to_subtract += episode.episode_shares.clone();
             }
         }
     });
 
+    if updated_rewards_at < current_time {
```

Comment:
Do you really need this check?
I would expect it will return 0 if updated_rewards_at == current_time